[versioning] EOS-25900: delete marker

ajay-paratmandali opened the pull request:
Signed-off-by: Ajay Paratmandali <ajay.paratmandali@seagate.com>

# Problem Statement
- If bucket is in suspended or Disabled state then follow old step
- If bucket is in enabled state then create delete marker metadata
- x-amz-delete-marker should only exists in metadata if delete_marker check is true

# Design
[-  For Bug, Describe the fix here.
-  For Feature, Post the link for design](https://seagate-systems.atlassian.net/l/c/psKmbxPU)

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

sjain09 commented on st/clitests/awss3api_spec.py:
```diff
@@ -2022,6 +2022,49 @@ def get_md5_sum(fname):
+#AwsTest('Aws can delete object in versioning state').delete_object("seagatebucket", "simpleobj1").execute_test().command_is_successful()
+
+#******** Suspend Versioning on Bucket ********
+AwsTest('Aws can suspend versioning on bucket').put_bucket_versioning("seagatebucket", "Suspended").execute_test().command_is_successful()
```
disable to buscket suspended case as well for now, as we don't support it for time being.

ajay-paratmandali replied:
updated

ajay-paratmandali replied:
Updated code
as there is no delete marker test now removed all st related to delete marker

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -308,6 +407,12 @@ void S3DeleteObjectAction::send_response_to_s3_client() {
     }
 
     request->send_response(error.get_http_status_code(), response_xml);
+  } else if (delete_marker_metadata) {
```
updated

keithpine commented on server/s3_object_metadata.h:
```diff
@@ -279,6 +280,10 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
 
+  // delete marker handler
+  virtual void set_delete_marker();
+  bool get_delete_marker();
```
```suggestion
  bool get_delete_marker() const;
```

ajay-paratmandali replied:
Updated

keithpine commented on server/s3_object_metadata.cc:
```diff
@@ -37,6 +37,18 @@
+    const struct s3_motr_idx_layout& obj_idx_lo,
+    const struct s3_motr_idx_layout& obj_ver_idx_lo) {
+
+  std::shared_ptr<S3ObjectMetadata> meta =
```
```suggestion
  auto meta =
```

`auto` is useful when the function makes it obvious what the type is.

ajay-paratmandali replied:
Updated

keithpine commented on server/s3_object_metadata.cc:
```diff
@@ -374,6 +386,15 @@ void S3ObjectMetadata::set_version_id(std::string ver_id) {
+  is_delete_marker = true;
+}
+
+bool S3ObjectMetadata::get_delete_marker() {
```
```suggestion
bool S3ObjectMetadata::get_delete_marker() const {
```

ajay-paratmandali replied:
Updated

keithpine commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  // TODO: update null version
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
+      s3_log(S3_LOG_DEBUG, request_id,
+             "Writing user metadata on object: [%s] -> [%s]\n",
+             it.first.c_str(), it.second.c_str());
+      delete_marker_metadata->add_user_defined_attribute(it.first, it.second);
+    }
+  }
```
I don't think I understand what's going on here. Can users set metadata on delete markers?

ajay-paratmandali replied:
I created interface in s3_factory to provide delete-marker metadata
```
  delete_marker_metadata =
      object_metadata_factory->create_delete_marker_metadata_obj(
          request, bucket_metadata->get_object_list_index_layout(),
          bucket_metadata->get_objects_version_list_index_layout());
```
which create metadata object and set delete marker before returning metadata

keithpine replied:
That doesn't really answer my question. Where is this user metadata coming from? A DeleteMarker is caused by a DeleteObject API request. I don't see any ability to set user metadata. The AWS CLI does not allow metadata. So the question is, why do you need to apply user metadata?

keithpine commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
```
Shouldn't it be:

```
    if (it.first.find("x-amz-meta-") == 0) {
```

Otherwise, it will find `x-amz-meta` anywhere in the string.

ajay-paratmandali replied:
Yes i was unaware of tis part
after testing
updated code

keithpine commented on server/s3_object_metadata.cc:
```diff
@@ -374,6 +386,13 @@ void S3ObjectMetadata::set_version_id(std::string ver_id) {
+  is_delete_marker = true;
+}
+
+bool S3ObjectMetadata::get_delete_marker() { return is_delete_marker; }
```
```suggestion
bool S3ObjectMetadata::get_delete_marker() const { return is_delete_marker; }
```

ajay-paratmandali replied:
Updated

keithpine commented on server/s3_object_metadata.cc:
```diff
@@ -930,6 +954,11 @@ int S3ObjectMetadata::from_json(std::string content) {
   rev_epoch_version_id_key =
       S3ObjectVersioingHelper::generate_keyid_from_versionid(object_version_id);
 
+  if (system_defined_attribute.find("x-amz-delete-marker") !=
+      system_defined_attribute.end()) {
+    is_delete_marker = true;
+  }
```
```suggestion
  is_delete_marker = system_defined_attribute.find("x-amz-delete-marker") !=
      system_defined_attribute.end();
```

ajay-paratmandali replied:
Updated

keithpine commented on st/clitests/awss3api_spec.py:
```diff
@@ -2022,6 +2022,43 @@ def get_md5_sum(fname):
+#******** Get Bucket Versioning Enabled status ********
+AwsTest('Aws can get bucket versioning Enabled status').get_bucket_versioning("seagatebucket").execute_test().command_response_should_have("Enabled")
+
+# TODO: uncomment in specific version delete as empty bucket not possible without this api.
+#******** Put Object in Bucket Versioning Suspended status ********
+#AwsTest('Aws can get bucket versioning Suspended status').get_bucket_versioning("seagatebucket").execute_test().command_response_should_have("Suspended")
+
+#******** Delete object in enable versioning state ********
+#AwsTest('Aws can delete object in versioning state').delete_object("seagatebucket", "simpleobj1").execute_test().command_is_successful()
```
Please don't add disabled tests in new code. Just implement them when needed.

ajay-paratmandali replied:
Removed Code

keithpine commented on server/s3_delete_object_action.cc:
```diff
@@ -76,10 +76,19 @@ void S3DeleteObjectAction::setup_steps() {
-  ACTION_TASK_ADD(S3DeleteObjectAction::populate_probable_dead_oid_list, this);
-  ACTION_TASK_ADD(S3DeleteObjectAction::delete_metadata, this);
+
+  /*
```
Please match the comment style of existing code, `//`.

ajay-paratmandali replied:
Updated comment

keithpine commented on server/s3_delete_object_action.cc:
```diff
@@ -76,10 +76,19 @@ void S3DeleteObjectAction::setup_steps() {
+
+  /*
+  delete handler
+    probable delete   // if bucket is suspended
+    create delete marker //if bucket is enabled
```
```suggestion
    probable delete   // if bucket versioning is suspended
    create delete marker // if bucket versioning is enabled
```

ajay-paratmandali replied:
Updated

keithpine commented on server/s3_delete_object_action.cc:
```diff
@@ -76,10 +76,19 @@ void S3DeleteObjectAction::setup_steps() {
+    probable delete   // if bucket is suspended
+    create delete marker //if bucket is enabled
+  object_metadata_handler
+    save delete_marker  // if bucket is suspended
+    delete_metadata //if bucket is enabled
```
```suggestion
    save delete_marker  // if bucket versioning is suspended
    delete_metadata // if bucket versioning is enabled
```

ajay-paratmandali replied:
Updated

ajay-paratmandali commented on ut/s3_delete_object_action_test.cc:
```diff
@@ -19,6 +19,8 @@
  */
 
 #include <memory>
+#include <map>
```
Corrected changes

ajay-paratmandali replied:
Updated

ajay-paratmandali commented on ut/s3_delete_object_action_test.cc:
```diff
@@ -422,3 +423,58 @@ TEST_F(S3DeleteObjectActionTest, DeleteObjectsDelayedEnabled) {
+
+  std::map<std::string, std::string> in_headers_copy;
+  in_headers_copy.insert(
+      std::pair<std::string, std::string>("x-amz-meta-a", "true"));
```
Updated

ajay-paratmandali commented on st/clitests/awss3api_spec.py:
```diff
@@ -2022,6 +2022,49 @@ def get_md5_sum(fname):
 AwsTest('Aws can delete bucket').delete_bucket("target-bucket")\
     .execute_test().command_is_successful()
 
+#******** Create Bucket ********
```
Removed st related code

ajay-paratmandali commented on st/clitests/awss3api_spec.py:
```diff
@@ -2022,6 +2022,49 @@ def get_md5_sum(fname):
+AwsTest('Aws can create bucket').create_bucket("seagatebucket").execute_test().command_is_successful()
+
+#******** Get Bucket Versioning default status ********
+AwsTest('Aws can get bucket versioning default status').get_bucket_versioning("seagatebucket").execute_test().command_response_should_be_empty()
```
Removed code as per above comment

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+
+  // TODO: update null version
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
```
This is not need log was created for testing 
Removed in next patch

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  // TODO: update null version
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
+  for (auto it : request->get_in_headers_copy()) {
```
Updated

sjain09 commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+void S3DeleteObjectAction::metadata_handler() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
```
Its an local copy.

ajay-paratmandali replied:
it is from parent class S3ObjectAction which is common code all other plugin like put object

sjain09 commented on server/s3_delete_object_action.h:
```diff
@@ -50,6 +52,9 @@ class S3DeleteObjectAction : public S3ObjectAction {
   std::shared_ptr<S3MotrKVSWriterFactory> mote_kv_writer_factory;
 
+  // delete marker metadata
+  std::shared_ptr<S3ObjectMetadata> delete_marker_metadata;
```
initialize `delete_marker_metadata` to nullptr in the constructor.

ajay-paratmandali commented on ut/mock_s3_factory.h:
```diff
@@ -77,6 +79,14 @@ class MockS3ObjectMetadataFactory : public S3ObjectMetadataFactory {
+      const struct s3_motr_idx_layout& obj_idx_lo = {},
+      const struct s3_motr_idx_layout& obj_ver_idx_lo = {}) override {
+
+    // mock_delete_marker_metadata->set_delete_marker();
```
uncommented as needed to set delete marker

missed this before 
updated now

sjain09 reviewed (approved):
Looks good.

keithpine reviewed (commented):
Some outstanding issues/questions:

1. What is the purpose of setting the user metadata on the DeleteMarker object? Is this necessary?
2. aws client fails when a DeleteMarker exists, because there is no `Size` value in the response.
3. s3server crashes when you try to delete a DeleteMarker by version ID (motr panic).

Given issues 2 and 3, should we wait to merge this until changes to handle those are ready to go in the same sprint?

sanalpk commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +132,90 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
```
is it required to cache? this could be (bucket_metadata->get_bucket_versioning_status() == "Enabled")

ajay-paratmandali replied:
Updated

sanalpk commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +132,90 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
```
same comment here as well

ajay-paratmandali replied:
Updated

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +132,90 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+             it.first.c_str(), it.second.c_str());
+      delete_marker_metadata->add_user_defined_attribute(it.first, it.second);
+    }
+  }
```
Removed this part as not needed

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -43,6 +43,7 @@ S3DeleteObjectAction::S3DeleteObjectAction(
          request->get_object_name().c_str());
 
   action_uses_cleanup = true;
+  delete_marker_metadata = nullptr;
```
Removed

ajay-paratmandali commented:
@keithpine 
>>>Some outstanding issues/questions:

What is the purpose of setting the user metadata on the DeleteMarker object? Is this necessary?
aws client fails when a DeleteMarker exists, because there is no Size value in the response.
s3server crashes when you try to delete a DeleteMarker by version ID (motr panic).
Given issues 2 and 3, should we wait to merge this until changes to handle those are ready to go in the same sprint?

What is the purpose of setting the user metadata on the DeleteMarker object? Is this necessary?
removed this part from code as not needed

aws client fails when a DeleteMarker exists, because there is no Size value in the response.
set size to zero not getting error now

s3server crashes when you try to delete a DeleteMarker by version ID (motr panic).
added OprationNotSupported error check

sjain09 commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +131,86 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  delete_marker_metadata->regenerate_version_id();
+  delete_marker_metadata->reset_date_time_to_current();
+  delete_marker_metadata->set_content_length(
+    request->get_data_length_str());
```
Hardcode it to zero.

ajay-paratmandali replied:
Updated

sjain09 reviewed (approved):
Looks Good.

sjain09 commented:
Create a separate jira for ST creation for future sprint.

keithpine commented:
Note the following behavior with this change:

1. DeleteObject with `versionId` now returns an error, until support is officially added. This means you cannot actually delete any objects from a versioned bucket for now.
2. Deleting the latest object inserts a DeleteMarker. A ListObjects command will show the delete marker as a normal object with a size of 0. This will be addressed in a follow up change.

ajay-paratmandali commented on server/s3_delete_object_action.cc:
```diff
@@ -125,6 +131,85 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
+  // Generate a version id for the new object.
+  delete_marker_metadata->regenerate_version_id();
+  delete_marker_metadata->reset_date_time_to_current();
+  delete_marker_metadata->set_content_length("0");
+  delete_marker_metadata->set_content_type(request->get_content_type());
```
it will be handled in delete specific version PR
with special case to check top is delete marker and then delete metadata but not data

ajay-paratmandali replied:
As per discussion will check on aws ls api and create new ticket if effort is more than one day
